cpp/src/strings/datetime.cu

```diff
@@ -0,0 +1,763 @@
+                    break;
+                }
+                case 'z':
+                {
+                    int sign = *ptr=='-' ? -1:1;
+                    int hh = str2int(ptr+1,2);
+                    int mm = str2int(ptr+3,2);
+                    // ignoring the rest for now
+                    // item.length has how many chars we should read
+                    timeparts[TP_TZ_MINUTES] = sign * ((hh*60)+mm);
+                    break;
+                }
+                default:
+                    return 3;
+            }
+            ptr += item.length;
+            length -= item.length;
+        }
+        return 0;
+    }
+
+    __device__ int64_t timestamp_from_parts( int32_t* timeparts, timestamp_units units )
+    {
+        auto year = timeparts[TP_YEAR];
+        if( units==timestamp_units::years )
+            return year - 1970;
+        auto month = timeparts[TP_MONTH];
+        if( units==timestamp_units::months )
+            return ((year-1970) * 12) + (month-1); // months are 1-12, need to 0-base it here
+        auto day = timeparts[TP_DAY];
+        // The months are shifted so that March is the starting month and February
+        // (possible leap day in it) is the last month for the linear calculation
+        year -= (month <= 2) ? 1 : 0;
+        // date cycle repeats every 400 years (era)
+        constexpr int32_t erasInDays = 146097;
+        constexpr int32_t erasInYears = (erasInDays / 365);
+        auto era = (year >= 0 ? year : year - 399) / erasInYears;
+        auto yoe = year - era * erasInYears;
+        auto doy = month==0 ? day : ((153 * (month + (month > 2 ? -3 : 9)) + 2) / 5 + day - 1);
+        auto doe = (yoe * 365) + (yoe / 4) - (yoe / 100) + doy;
+        int32_t days = (era * erasInDays) + doe - 719468; // 719468 = days from 0000-00-00 to 1970-03-01
+        if( units==timestamp_units::days )
+            return days;
+
+        auto tzadjust = timeparts[TP_TZ_MINUTES]; // in minutes
+        auto hour = timeparts[TP_HOUR];
+        if( units==timestamp_units::hours )
+            return (days*24L) + hour + (tzadjust/60);
+
+        auto minute = timeparts[TP_MINUTE];
+        if( units==timestamp_units::minutes )
+            return static_cast<int64_t>(days * 24L * 60L) + (hour * 60L) + minute + tzadjust;
+
+        auto second = timeparts[TP_SECOND];
+        int64_t timestamp = (days * 24L * 3600L) + (hour * 3600L) + (minute * 60L) + second + (tzadjust*60);
+        if( units==timestamp_units::seconds )
+            return timestamp;
+
+        auto subsecond = timeparts[TP_SUBSECOND];
+        if( units==timestamp_units::ms )
+            timestamp *= 1000L;
+        else if( units==timestamp_units::us )
+            timestamp *= 1000000L;
+        else if( units==timestamp_units::ns )
+            timestamp *= 1000000000L;
+        timestamp += subsecond;
+        return timestamp;
+    }
+
+     __device__ T operator()(size_type idx)
+    {
+        if( d_strings.is_null(idx) )
```

Comment:
A common pattern to avoid this check when `d_strings` is non-nullable is to add a `has_nulls` template bool. 

For example: https://github.com/rapidsai/cudf/blob/branch-0.11/cpp/include/cudf/table/row_operators.cuh#L146

Reply:
> A common pattern to avoid this check when `d_strings` is non-nullable is to add a `has_nulls` template bool.
> 

Is this a requirement before approving this?

Reply:
@jrhemstad ?

Reply:
No.